File: src/main/java/org/vors/pairbot/service/CommandService.java
package org.vors.pairbot.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
import org.telegram.telegrambots.meta.api.objects.EntityType;
import org.telegram.telegrambots.meta.api.objects.Message;
import org.telegram.telegrambots.meta.api.objects.MessageEntity;
import org.telegram.telegrambots.meta.bots.AbsSender;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;
import org.vors.pairbot.generator.PairGenerator;
import org.vors.pairbot.model.Event;
import org.vors.pairbot.model.Team;
import org.vors.pairbot.model.UserInfo;
import org.vors.pairbot.repository.EventRepository;
import org.vors.pairbot.repository.TeamRepository;
import org.vors.pairbot.repository.UserRepository;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static org.vors.pairbot.constant.BotCommands.*;


@Transactional
@Component
public class CommandService {
    private Logger LOG = LoggerFactory.getLogger(CommandService.class);

    @Autowired
    @Lazy
    private AbsSender bot;
    @Autowired
    private UserService userService;

    @Autowired
    private MessageService messageService;
    @Autowired
    private KeyboardService keyboardService;
    @Autowired
    private PairGenerator pairGenerator;
    @Autowired
    private UserRepository userRepository;
    @Autowired
    private ChatService chatService;
    @Autowired
    private EventRepository eventRepository;
    @Autowired
    private TeamRepository teamRepository;


    public void processCommand(Message message) throws TelegramApiException {
        Optional<String> commandTextOpt = extractCommandText(message);

        if (commandTextOpt.isPresent()) {
            String commandText = commandTextOpt.get();

            Long chatId = message.getChatId();
            UserInfo user = userService.findByUserId(message.getFrom().getId()).get();
            switch (commandText) {
                case START:
                    String messageText = message.getText();
                    if (commandText.length() < messageText.length()) {
                        String teamToken = messageText.substring(commandText.length() + 1);
                        joinTeamByToken(teamToken, user);

                        messageService.sendMessage(chatId, messageService.teamInfo(user));

                    } else {
                        SendMessage sendMessage = messageService.getMessageWithKeyboard(
                                chatId,
                                "Hi! Let's make a team. \nOr ask your peers for a link to join",
                                keyboardService.getStartKeyboard());
                        messageService.sendMessage(sendMessage);
                    }
                    break;
                case PAIR:
                    if (pairGenerator.canCreateEvent(user)) {
                        createEventAndInvite(user);
                    } else {
                        messageService.sendMessage(chatId, messageService.tryLaterText(user));
                    }
                    break;
                case MY_TEAM:
                    messageService.sendMessage(chatId, messageService.teamInfo(user));
                    break;
                case SET_LOCATION:
                    messageService.requestLocation(chatId);
                    break;

            }
        }
    }

    private void joinTeamByToken(String token, UserInfo user) {
        Optional<Team> teamOpt = teamRepository.findByToken(UUID.fromString(token));

        teamOpt.ifPresent(team -> {
            team.addMember(user);
            teamRepository.save(team);
        });
    }

    private void createEventAndInvite(UserInfo user) throws TelegramApiException {
        Optional<Event> eventOpt = pairGenerator.findPair(user);

        if (eventOpt.isPresent()) {
            Event event = eventOpt.get();
            eventRepository.save(event);

            messageService.sendToAll(event, messageService::inviteText, keyboardService::getInviteKeyboard);

        } else {
            sendNoPairsAvailable(user);
        }
    }



    private void sendNoPairsAvailable(UserInfo user) throws TelegramApiException {
        SendMessage message = messageService.getMessage(
                chatService.getPrivateChatId(user),
                "No pairs available.\nEveryone is busy or no one joined yet \uD83D\uDE1E"
        );
        messageService.sendMessage(message);
    }

    private Optional<String> extractCommandText(Message message) {
        Optional<List<MessageEntity>> entitiesOpt = Optional.ofNullable(message.getEntities());

        return entitiesOpt.flatMap(this::extractCommandText);
    }

    private Optional<String> extractCommandText(List<MessageEntity> entities) {
        return entities.stream()
                .filter(e -> e != null && e.getOffset() == 0 && EntityType.BOTCOMMAND.equals(e.getType()))
                .findFirst()
                .map(MessageEntity::getText)
                .map(this::removeMention);
    }

    private String removeMention(String commandText) {
        int indexOfMention = commandText.indexOf("@");
        if (indexOfMention != -1) {
            commandText = commandText.substring(0, indexOfMention);
        }
        return commandText;
    }

}

File: src/main/java/org/vors/pairbot/service/CommandService.kt
package org.vors.pairbot.service

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import org.telegram.telegrambots.meta.api.methods.ParseMode
import org.telegram.telegrambots.meta.api.objects.EntityType
import org.telegram.telegrambots.meta.api.objects.Message
import org.telegram.telegrambots.meta.api.objects.MessageEntity
import org.telegram.telegrambots.meta.exceptions.TelegramApiException
import org.vors.pairbot.constant.BotCommand.*
import org.vors.pairbot.event.EventOrganizer
import org.vors.pairbot.model.UserInfo
import org.vors.pairbot.repository.TeamRepository
import java.util.*


@Transactional
@Component
open class CommandService(
        open var userService: UserService,
        open var messageService: MessageService,
        open var keyboardService: KeyboardService,
        open var chatService: ChatService,
        open var teamRepository: TeamRepository,
        open var eventOrganizer: EventOrganizer
) {
    private val LOG = LoggerFactory.getLogger(CommandService::class.java)


    @Throws(TelegramApiException::class)
    fun processCommand(message: Message) {
        val commandTextOpt = extractCommandText(message)

        if (commandTextOpt.isPresent) {
            val commandText = commandTextOpt.get()

            val chatId = message.chatId
            val user = userService.findByUserId(message.from.id).get()
            when (valueOf(commandText.toUpperCase())) {
                START -> {
                    val messageText = message.text
                    if (commandText.length < messageText.length - 1) {
                        val teamToken = messageText.substring(commandText.length + 1).trim()
                        joinTeamByToken(teamToken, user)

                        messageService.sendMessage(chatId, messageService.teamInfo(user))

                    } else {
                        val sendMessage = messageService.getMessageWithKeyboard(
                                chatId,
                                "Hi! Let's make a team. \nOr ask your peers for a link to join",
                                keyboardService.startKeyboard,
                                ParseMode.MARKDOWN)
                        messageService.sendMessage(sendMessage)
                    }
                }
                PAIR -> {
                    eventOrganizer.tryOrganizeEvent(user)
                }
                MYTEAM, TEAM -> messageService.sendMessage(chatId, messageService.teamInfo(user))
                SET_LOCATION -> messageService.requestLocation(chatId)
                VOID -> TODO()
            }
        }
    }

    private fun joinTeamByToken(token: String, user: UserInfo) {
        val teamOpt = teamRepository.findByToken(UUID.fromString(token))

        teamOpt.ifPresent { team ->
            team.addMember(user)
            teamRepository.save(team)
        }
    }

    private fun extractCommandText(message: Message): Optional<String> {
        val entitiesOpt = Optional.ofNullable(message.entities)

        return entitiesOpt.flatMap(this::extractCommandText)
    }

    private fun extractCommandText(entities: List<MessageEntity>): Optional<String> {
        return entities.stream()
                .filter { e -> e != null && e.offset == 0 && EntityType.BOTCOMMAND == e.type }
                .findFirst()
                .map { this.removeMentionAndSlash(it.text) }
    }

    private fun removeMentionAndSlash(commandText: String): String {

        val startPos = if (commandText[0] == '/') 1 else 0

        val indexOfMention = commandText.indexOf("@")

        return if (indexOfMention != -1) {
            commandText.substring(startPos, indexOfMention)
        } else {
            commandText.substring(startPos)
        }
    }

}
